core/domain/voiceover_domain.py

```diff
@@ -104,101 +112,97 @@ def to_dict(self) -> EntityVoiceoverDict:
         )
 
     def validate(self) -> None:
-        """Validates the EntityVoiceover object."""
+        """Validates the EntityVoiceovers object."""
         if not isinstance(self.entity_type, str):
             raise utils.ValidationError(
-                'entity_type must be a string, recieved %r' % self.entity_type)
+                'entity_type must be a string, received %s' % self.entity_type)
         if not isinstance(self.entity_id, str):
             raise utils.ValidationError(
-                'entity_id must be a string, recieved %r' % self.entity_id)
+                'entity_id must be a string, received %s' % self.entity_id)
         if not isinstance(self.entity_version, int):
             raise utils.ValidationError(
-                'entity_version must be an int, recieved %r' %
+                'entity_version must be an int, received %s' %
                 self.entity_version)
         if not isinstance(self.language_accent_code, str):
             raise utils.ValidationError(
-                'language_accent_code must be a string, recieved %r' %
+                'language_accent_code must be a string, received %s' %
                 self.language_accent_code)
-
+        if not bool(re.match(
+                feconf.LANGUAGE_ACCENT_CODE_REGEX, self.language_accent_code)):
+            raise utils.ValidationError(
+                'language_accent_code must follow the pattern language-accent'
+                ' code, received %s' % self.language_accent_code)
         for content_id, voiceover_type_to_voiceover in self.voiceovers.items():
             if not isinstance(content_id, str):
                 raise utils.ValidationError(
-                    'content_id must be a string, recieved %r' % content_id)
+                    'content_id must be a string, received %s' % content_id)
             for voiceover_type, voiceover in (
                     voiceover_type_to_voiceover.items()):
-                if not isinstance(voiceover_type, str):
+                if not isinstance(voiceover_type, feconf.VoiceoverType):
                     raise utils.ValidationError(
-                        'voiceover type must be string, recieved %r' %
+                        'voiceover type must be VoiceoverType, received %s' %
                         voiceover_type)
-                if voiceover_type not in ['manual', 'auto']:
-                    raise utils.ValidationError(
-                        'The voiceover type must be either manual or auto, but '
-                        'received %r' % voiceover_type)
                 voiceover.validate()
 
     def add_voiceover(
         self,
         content_id: str,
-        voiceover_type: str,
+        voiceover_type: feconf.VoiceoverType,
         voiceover: state_domain.Voiceover
     ) -> None:
         """Adds voiceover to the entity voiceover instance."""
-        if not isinstance(content_id, str):
+        if not isinstance(voiceover_type, feconf.VoiceoverType):
```

Comment:
Shouldn't this (and other type stuff) be checked in the controller layer? That and validate() should probably be the only places.

Reply:
Removed